File: chapters/04-probability-basics/c/dice_and_distributions.c
/*
 * Chapter 4 - probability verified by brute force, in pure C.
 *
 * Full port of the Python example: two-dice distribution by simulation vs
 * exact counting, expected values by sampling, and cross-entropy grading of
 * two weather forecasters.
 *
 * Note the fixed seed (srand(42)): every reader sees the same "random"
 * numbers, so the printed output is reproducible. Seeding is a habit kept
 * through the whole course - it is how you debug anything involving
 * randomness.
 *
 * Build and run from the repository root:
 *     make -C chapters/04-probability-basics/c
 *     ./chapters/04-probability-basics/c/build/dice_and_distributions
 */

#include <math.h>
#include <stdio.h>
#include <stdlib.h>

/*
 * Return one fair die roll, an integer from 1 to 6.
 *
 * rand() returns 0..RAND_MAX; the modulo folds it into 0..5 and the +1 shifts
 * to 1..6. The tiny bias this introduces (RAND_MAX is not a multiple of 6) is
 * far below what a million rolls can detect, so it is fine for this chapter.
 */
static int roll_one_die(void) {
    return rand() % 6 + 1;
}

/*
 * Return the exact probability of a given two-dice sum by counting pairs.
 *
 * dice_sum: the sum in question, 2 through 12.
 *
 * Out of 36 equally likely pairs, the number summing to s is 6 - |7 - s|.
 */
static double exact_two_dice_probability(int dice_sum) {
    int number_of_pairs = 6 - abs(7 - dice_sum);
    return number_of_pairs / 36.0;
}

/*
 * Compute a forecaster's cross-entropy: their average surprise at reality.
 *
 * rain_probabilities_given:  probability the forecaster gave to "rain", per day
 * it_actually_rained:        1 where it rained, 0 where it did not, per day
 * number_of_days:            how many days both arrays cover
 *
 * Returns the average of -log(probability given to what actually happened).
 * On a dry day the probability given to the actual outcome is 1 - p_rain,
 * which is why both branches appear below.
 */
static double compute_cross_entropy_for_rain_forecasts(const double *rain_probabilities_given,
                                                       const int *it_actually_rained,
                                                       int number_of_days) {
    double total_surprise = 0.0;
    for (int day_index = 0; day_index < number_of_days; day_index++) {
        double probability_given_to_actual_outcome =
            it_actually_rained[day_index] ? rain_probabilities_given[day_index]
                                          : 1.0 - rain_probabilities_given[day_index];
        total_surprise += -log(probability_given_to_actual_outcome);
    }
    return total_surprise / number_of_days;
}

int main(void) {
    srand(42);

    const int number_of_rolls = 1000000;
    printf("1. Two-dice distribution: %d simulated rolls vs exact counting\n", number_of_rolls);

    long sum_counts[13] = {0};
    for (int roll_index = 0; roll_index < number_of_rolls; roll_index++) {
        sum_counts[roll_one_die() + roll_one_die()]++;
    }
    printf("  sum   simulated   exact\n");
    for (int dice_sum = 2; dice_sum <= 12; dice_sum++) {
        printf("  %3d   %.5f     %.5f\n", dice_sum,
               (double)sum_counts[dice_sum] / number_of_rolls,
               exact_two_dice_probability(dice_sum));
    }

    printf("\n2. Expected values estimated by sampling (exact: 3.5 and 7.0)\n");
    const int sample_sizes[] = {100, 10000, 1000000};
    for (int size_index = 0; size_index < 3; size_index++) {
        int number_of_samples = sample_sizes[size_index];
        long single_die_total = 0;
        long two_dice_total = 0;
        for (int sample_index = 0; sample_index < number_of_samples; sample_index++) {
            single_die_total += roll_one_die();
            two_dice_total += roll_one_die() + roll_one_die();
        }
        printf("  %9d samples: one die = %.4f, two dice = %.4f\n", number_of_samples,
               (double)single_die_total / number_of_samples,
               (double)two_dice_total / number_of_samples);
    }

    printf("\n3. Cross-entropy: grading two weather forecasters (it rained on days 1, 2, 5)\n");
    const int it_actually_rained[] = {1, 1, 0, 0, 1};
    const double forecaster_a_probabilities[] = {0.8, 0.9, 0.2, 0.1, 0.6};
    const double forecaster_b_probabilities[] = {0.5, 0.5, 0.5, 0.5, 0.5};
    printf("  forecaster A (confident and mostly right): %.2f\n",
           compute_cross_entropy_for_rain_forecasts(forecaster_a_probabilities, it_actually_rained, 5));
    printf("  forecaster B (always says 50/50):          %.2f\n",
           compute_cross_entropy_for_rain_forecasts(forecaster_b_probabilities, it_actually_rained, 5));
    printf("  Lower is better: cross-entropy rewards confident correctness.\n");
    return 0;
}

File: chapters/04-probability-basics/python/dice_and_distributions.py
"""Chapter 4 - probability verified by brute force: dice, expectation, cross-entropy.

Three demonstrations:
  1. Simulate a million two-dice rolls; the empirical distribution matches the
     exact counting argument (law of large numbers).
  2. Sample averages converge to the expected values (3.5 for one die).
  3. Score two weather forecasters with cross-entropy, reproducing the
     chapter's hand-computed table.

Run from the repository root:
    .venv/bin/python chapters/04-probability-basics/python/dice_and_distributions.py
"""

import math
import random


def simulate_two_dice_distribution(number_of_rolls, random_generator):
    """Roll two dice many times and count how often each sum appears.

    Arguments:
        number_of_rolls: how many times to roll the pair of dice.
        random_generator: a random.Random instance. Passing it in (instead of
            using the global one) keeps every run reproducible from one seed.

    Returns a dict mapping each sum (2..12) to the fraction of rolls that
    produced it.
    """
    sum_counts = {possible_sum: 0 for possible_sum in range(2, 13)}
    for _ in range(number_of_rolls):
        first_die = random_generator.randint(1, 6)
        second_die = random_generator.randint(1, 6)
        sum_counts[first_die + second_die] += 1
    return {dice_sum: count / number_of_rolls for dice_sum, count in sum_counts.items()}


def exact_two_dice_probability(dice_sum):
    """Return the exact probability of a given two-dice sum by counting pairs.

    Arguments:
        dice_sum: the sum in question, 2 through 12.

    Out of 36 equally likely pairs, the number summing to s is 6 - |7 - s|
    (1 pair for 2 or 12, rising to 6 pairs for 7).
    """
    number_of_pairs = 6 - abs(7 - dice_sum)
    return number_of_pairs / 36.0


def estimate_expected_value_by_sampling(number_of_samples, random_generator):
    """Estimate E[one die] and E[sum of two dice] by averaging samples.

    Arguments:
        number_of_samples: how many rolls to average.
        random_generator: a random.Random instance.

    Returns the pair (average of one die, average of two-dice sums).
    """
    single_die_total = 0
    two_dice_total = 0
    for _ in range(number_of_samples):
        single_die_total += random_generator.randint(1, 6)
        two_dice_total += random_generator.randint(1, 6) + random_generator.randint(1, 6)
    return single_die_total / number_of_samples, two_dice_total / number_of_samples


def compute_cross_entropy_for_rain_forecasts(rain_probabilities_given, it_actually_rained):
    """Compute a forecaster's cross-entropy: their average surprise at reality.

    Arguments:
        rain_probabilities_given: list of probabilities the forecaster gave to
            "it will rain", one per day.
        it_actually_rained: list of booleans, one per day, True where it rained.

    Returns the average of -log(probability given to what actually happened).
    On a day without rain, the probability given to the actual outcome is
    1 - (rain probability), which is why both branches appear below.
    """
    total_surprise = 0.0
    for rain_probability, rained in zip(rain_probabilities_given, it_actually_rained):
        probability_given_to_actual_outcome = rain_probability if rained else 1.0 - rain_probability
        total_surprise += -math.log(probability_given_to_actual_outcome)
    return total_surprise / len(rain_probabilities_given)


def main():
    random_generator = random.Random(42)

    number_of_rolls = 1_000_000
    print(f"1. Two-dice distribution: {number_of_rolls:,} simulated rolls vs exact counting")
    empirical_distribution = simulate_two_dice_distribution(number_of_rolls, random_generator)
    print("  sum   simulated   exact")
    for dice_sum in range(2, 13):
        print(f"  {dice_sum:>3}   {empirical_distribution[dice_sum]:.5f}     {exact_two_dice_probability(dice_sum):.5f}")

    print()
    print("2. Expected values estimated by sampling (exact: 3.5 and 7.0)")
    for number_of_samples in (100, 10_000, 1_000_000):
        single_die_average, two_dice_average = estimate_expected_value_by_sampling(
            number_of_samples, random_generator
        )
        print(f"  {number_of_samples:>9,} samples: one die = {single_die_average:.4f}, two dice = {two_dice_average:.4f}")

    print()
    print("3. Cross-entropy: grading two weather forecasters (it rained on days 1, 2, 5)")
    it_actually_rained = [True, True, False, False, True]
    forecaster_a_probabilities = [0.8, 0.9, 0.2, 0.1, 0.6]
    forecaster_b_probabilities = [0.5, 0.5, 0.5, 0.5, 0.5]
    cross_entropy_a = compute_cross_entropy_for_rain_forecasts(forecaster_a_probabilities, it_actually_rained)
    cross_entropy_b = compute_cross_entropy_for_rain_forecasts(forecaster_b_probabilities, it_actually_rained)
    print(f"  forecaster A (confident and mostly right): {cross_entropy_a:.2f}")
    print(f"  forecaster B (always says 50/50):          {cross_entropy_b:.2f}")
    print("  Lower is better: cross-entropy rewards confident correctness.")


if __name__ == "__main__":
    main()
